dacbench/envs/layerwise_sgd.py

```diff
@@ -0,0 +1,539 @@
+                ]
+            )
+
+            local_observations_tensor = torch.cat(local_observations)
+            layer_state = torch.cat([local_observations_tensor, global_observations_tensor])
+            states.append(layer_state)
+
+        return states
+
+    def render(self, mode="human"):
+        """Render progress."""
+        if mode == "human":
+            epoch = 1 + self.c_step // len(self.train_loader)
+            epoch_cutoff = self.n_steps // len(self.train_loader)
+            batch = 1 + self.c_step % len(self.train_loader)
+            print(
+                f"prev_lr {self.optimizer.param_groups[0]['lr'] if self.n_steps > 0 else None}, "  # noqa: E501
+                f"epoch {epoch}/{epoch_cutoff}, "
+                f"batch {batch}/{len(self.train_loader)}, "
+                f"batch_loss {self.train_loss}, "
+                f"val_loss {self.validation_loss}"
+            )
+        else:
+            raise NotImplementedError
+
+    def forward_backward(self, model, loss_function, loader, device="cpu"):
+        """Do a forward and a backward pass for given `model` for `loss_function`.
+
+        Returns:
+            loss: Mini batch training loss per data point
+        """
+        model.train()
+        (data, target) = next(iter(loader))
+        data, target = data.to(device), target.to(device)
+        output = model(data)
+        loss = loss_function(output, target)
+        loss.mean().backward()
+
+        # for comparing current and last predictions
+        self.predictions.pop()
+        self.predictions.appendleft(output)
+
+        accuracy = torch.sum(output.argmax(dim=1) == target) / len(target)
+        return loss.mean().detach().cpu().numpy(), torch.tensor(accuracy).cpu().numpy()
+
+    def run_epoch(self, model, loss_function, loader, optimizer, device="cpu"):
+        """Run a single epoch of training for given `model` with `loss_function`."""
+        last_loss = None
+        running_loss = 0
+        predictions = []
+        model.train()
+        for data, target in loader:
+            data, target = data.to(device), target.to(device)
+            optimizer.zero_grad()
+            output = model(data)
+            loss = loss_function(output, target)
+            loss.mean().backward()
+            optimizer.step()
+            predictions.append(output)
+            last_loss = loss
+            running_loss += last_loss.mean().item()
+        self.predictions.pop()
+        self.predictions.appendleft(predictions.mean())
+        return last_loss.mean().detach().cpu(), running_loss.cpu() / len(loader)
+
+    def _update_lr_histories(self, log_learning_rates: list[float]) -> None:
+        for lr_history, log_lr in zip(self.lr_histories, log_learning_rates):
+            lr_history.pop()
+            lr_history.appendleft(log_lr)
+
+    def seed(self, seed, seed_action_space=False):
+        super(LayerwiseSGDEnv, self).seed(seed, seed_action_space)
```

Comment:
Just out of correctness (since i just read about it [here](https://stackoverflow.com/a/18208725)), generally it is bad practice to call super(__class__, self), due to the risk of Recursion exceptions. AFAIK its because class B can be not child but actually grandchild or deeper of class A.